src/suas_planning/src/GlobalWaypointPlanner.cpp

```diff
@@ -1,28 +1,141 @@
+        }
+    }
+    sub_goals.push(goal_);
+    // Start iterating through the sub_goals and execute a from waypoints
+    while (sub_goals.size() > 0) {
+        // Get first item goal off and pop goal off stack
+        Waypoint current_goal = sub_goals.front();
+        sub_goals.pop();
+        auto goal = AStarSearch(current_start, current_goal);
+    }
+
+    return {""};
 }
 
+/*
+ * Implements offline A* given start and sub goal
+ * Note due to the use of a unordered set with a priority queue as the frontier (probably should just subclass this);
+ * You need to first do operations on the pq_frontier unordered_set before pq (assuming you need to use std::move)
+ */
+std::shared_ptr<GlobalWaypointPlanner::Node> GlobalWaypointPlanner::AStarSearch(Waypoint start, Waypoint goal) {
+    // lambda function to compare Node objects
+    auto node_compare = [] (const std::shared_ptr<Node>& a, const::std::shared_ptr<Node>& b) { return *a < *b; };
+    // Creates a Min Heap for A*/Shortest Path Search
+    std::priority_queue<
+        std::shared_ptr<Node>,
+        std::vector<std::shared_ptr<Node>>,
+        decltype(node_compare)> pq(node_compare);
+    std::unordered_set<Node, std::hash<Node>, std::equal_to<Node>> pq_frontier; // just a hack so we check the frontier for nodes ()
+    std::unordered_set<Node, std::hash<Node>, std::equal_to<Node>> explored;
+
+    // create node using starting waypoint and push that onto related data sturcutres
+    auto current = std::make_shared<Node>(start_.x_, start.y_, goal, NO_ACTION, nullptr, *this);
+    pq_frontier.insert(*current);
+    pq.push(std::move(current));
+    while (pq.size()) {
+        // Pop the current node off of PQ and Frontier
+        std::shared_ptr<Node> current_node = pq.top();
+        pq_frontier.erase(*current_node);
+        pq.pop();
+
+        // Insert into explored
+        explored.insert(*current_node);
+        for (int action = 0; action < NUM_ACTIONS; action++) {
+            // Expand children for current node
+            int new_x = current_node->x_ + dX[action];
+            int new_y = current_node->y_ + dY[action];
+            bool in_bounds = Obstacle::CheckGridBounds(current_map_, new_x, new_y, map_meta_);
+            bool is_free = Obstacle::IsFree(current_map_, new_x, new_y, map_meta_);
+            if (in_bounds && is_free) {
+                auto child_node = std::make_shared<Node>(new_x, new_y, goal, action, current, *this);
+                if (explored.find(*child_node) != explored.end() ||
+                    pq_frontier.find(*child_node) != pq_frontier.end()) {
+                    if (child_node->IsGoal()) {
+                        // Found a solution, return it back up for processing
+                        return child_node;
+                    }
+                    pq_frontier.insert(*child_node);
+                    pq.push(std::move(child_node));
+                }
+            }
+        }
+    }
+
+    return nullptr;
+}
+
+
```

Comment:
Couldn't put a comment for line 159, but looking at how it it structured it looks like a good place to apply a standard algorithm. You'll have to look at examples/the docs for this, but it looks like a good place to use std::min_element and std::max_element or std::minmax (all with a custom comparator (that calls ::GetMinX and ::GetMaxX). When/if you merge ComputeMapWidth and ComputeMapHeight, you could still combine the two using std::minmax. These algorithms do essentially the same exact thing your for loop does, but more concise. std::min/max_element/std::minmax under the hood essentially runs this same for loop.

Reply:
Only small grip here is that the obstacles vector does not contain both the goal and start points. Would it be still worth refactoring to use min_element. 

Just realized that this code does not consider that any the waypoints might actually also live outside of that space as well, so lets throw that in as well: the start, goal, obstacles and waypoints.

Reply:
I wound up making the changes anyway. Refer to commit 8277de0